Fix randomly failures when deploying to OpenShift/K8s on JVM/Native

Sgitario opened the pull request:
Finally, I've found out the root cause of this issue. There were several places where a new KubernetesClient was being created:
- Some places use `Clients.fromConfig`.
- Other places use `KubernetesClientUtils.createConfig`.
- And other places use `KubernetesClientBuildItem.getClient`

And the problem is that every time we invoke any of these methods, Fabric8 KubernetesClient tries to locate one HttpClient.Factory and it seems that the logic to get one HttpClient.Factory sometimes gets the `VertxHttpClientFactory` implementation over the expected one `QuarkusHttpClientFactory`.

The previous solution about keeping the Disable Http Dns system property worked fine because this property was used either by `VertxHttpClientFactory` or `QuarkusHttpClientFactory`.

However, I've updated the pull request with a better solution and more efficient one as it avoids finding a HttpClient.Factory via the ServiceLoader logic in Fabric8 Kubernetes Client, but instead it directly provides the expected `QuarkusHttpClientFactory` implementation always.

Also, the warning message that always appeared, it's now gone:

```
[WARNING] [io.netty.resolver.dns.DefaultDnsServerAddressStreamProvider] Default DNS servers: [/[2001:4860:4860:0:0:0:0:8888]:53, /[2001:4860:4860:0:0:0:0:8844]:53] (Google Public DNS as a fallback)
[WARNING] There are multiple httpclient implementation in the classpath, choosing the first non-default implementation. You should exclude dependencies that aren't needed or use an explicit association of the HttpClient.Factory.
```

Fix https://github.com/quarkusio/quarkus/issues/31476

Sgitario commented:
@yrodiere with these changes, it seems to be consistently working for me. Could you try it yourself too just to be sure?

yrodiere commented on extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KnativeDeployer.java:
```diff
@@ -16,6 +17,8 @@ public class KnativeDeployer {
 
     @BuildStep
     public void checkEnvironment(Optional<SelectedKubernetesDeploymentTargetBuildItem> selectedDeploymentTarget,
+            // this is required to ensure the kubernetes client is fully configured when calling the `client.isSupported()`:
+            KubernetesClientOriginalValueDisableDnsResolverBuildItem kubernetesClientConfigured,
```
FYI you can also use `@Consume`:

```suggestion
    @BuildStep
    // this is required to ensure the kubernetes client is fully configured when calling the `client.isSupported()`:
    @Consume(KubernetesClientOriginalValueDisableDnsResolverBuildItem.class)
    public void checkEnvironment(Optional<SelectedKubernetesDeploymentTargetBuildItem> selectedDeploymentTarget,
```

... also, you will need to add an import.

Sgitario replied:
I've changed the solution, this change does not apply any longer.

yrodiere commented on extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/OpenshiftDeployer.java:
```diff
@@ -17,6 +18,8 @@ public class OpenshiftDeployer {
 
     @BuildStep
     public void checkEnvironment(Optional<SelectedKubernetesDeploymentTargetBuildItem> selectedDeploymentTarget,
+            // this is required to ensure the kubernetes client is fully configured when calling the `client.isSupported()`:
+            KubernetesClientOriginalValueDisableDnsResolverBuildItem kubernetesClientConfigured,
```
FYI you can also use `@Consume`:

```suggestion
    @BuildStep
    // this is required to ensure the kubernetes client is fully configured when calling the `client.isSupported()`:
    @Consume(KubernetesClientOriginalValueDisableDnsResolverBuildItem.class)
    public void checkEnvironment(Optional<SelectedKubernetesDeploymentTargetBuildItem> selectedDeploymentTarget,
```

... also, you will need to add an import.

Sgitario replied:
I've changed the solution, this change does not apply any longer.

geoand commented:
Can you please explain this change. I don't understand it

Sgitario commented:
> Can you please explain this change. I don't understand it

This is based on my findings here: https://github.com/quarkusio/quarkus/issues/31476#issuecomment-1449775877

yrodiere reviewed (commented):
I can't easily test with a Quarkus snapshot as I've only ever deployed from CI, but @gsmet is giving a try with another app that he usually deploys from his laptop, so let's wait for that.

As to the solution, I'm not sure I understand the content, but I added a comment about the form :)

geoand commented:
> > Can you please explain this change. I don't understand it
> 
> This is based on my findings here: [#31476 (comment)](https://github.com/quarkusio/quarkus/issues/31476#issuecomment-1449775877)

I'd like a more complete explanation please. Both to help me understand why this change fixes the problem and to have an easily accessible record of why this change was made.

Sgitario commented:
> > > Can you please explain this change. I don't understand it
> > 
> > 
> > This is based on my findings here: [#31476 (comment)](https://github.com/quarkusio/quarkus/issues/31476#issuecomment-1449775877)
> 
> I'd like a more complete explanation please. Both to help me understand why this change fixes the problem and to have an easily accessible record of why this change was made.

My guess was that the system property was cleared up and a new vert.x instance was created with the wrong value. Anyway, @gsmet is getting a different error after trying these changes. So, I'm changing this pull request to draft and keep working on it.

geoand commented:
> So, I'm changing this pull request to draft and keep working on it.

👍🏼.

Let me know if you need me to look into anything

Sgitario commented:
@geoand I've just updated the pull request, more information about the changes is in the description.
@yrodiere I've tried these changes using the GitHub lottery repo and it seems to be consistently working fine either in JVM and Native.

Sgitario commented:
@gsmet I've also tried to deploy the (quarkus-github-bot)[https://github.com/quarkusio/quarkus-github-bot] repo into OpenShift and it's working fine too (no warnings messages are printed).

@gsmet @yrodiere I would like to try these changes yourself too just to be surer that all the issues are gone.

Sgitario commented:
there is a regression with these changes that cause an out-of-memory error. I can reproduce it locally, investigating.

geoand commented:
That almost certainly means that there are multiple Vert.x instances

Sgitario commented:
It should be fixed with the latest changes.

gsmet commented:
I will test it with the Quarkus Bot and the production cluster this afternoon.

gsmet commented:
So... I don't think it's the same issue but I still have an error when pushing the Quarkus bot to the production cluster (error which apparently doesn't prevent the deployment as I have a success in the end):

```
[INFO] [io.quarkus.container.image.openshift.deployment.OpenshiftProcessor] Applied: ImageStream ubi-quarkus-native-binary-s2i
[INFO] [io.quarkus.container.image.openshift.deployment.OpenshiftProcessor] Applied: ImageStream quarkus-bot
[INFO] [io.quarkus.container.image.openshift.deployment.OpenshiftProcessor] Applied: BuildConfig quarkus-bot
[ERROR] Failed to upload archive file for the build: quarkus-bot
[ERROR] Please check cluster events via `oc get events` to see what could have possibly gone wrong
[WARNING] [io.quarkus.container.image.openshift.deployment.OpenshiftProcessor] An exception: 'Can't instantiate binary build, due to error reading/writing stream. Can be caused if the output stream was closed by the server.See if something's wrong in recent events in Cluster = Scheduled quarkus-bot-1-build.17489735390326d6 Successfully assigned prod-quarkus-bot/quarkus-bot-1-build to ip-10-0-142-232.us-west-2.compute.internal by ip-10-0-182-138
Created quarkus-bot-1-build.17489735b5f74d2d Created container git-clone
Started quarkus-bot-1-build.17489735b86fbe3a Started container git-clone
BuildStarted quarkus-bot-1.17489735e85618ad Build prod-quarkus-bot/quarkus-bot-1 is now running
AddedInterface quarkus-bot-1-build.17489735a33f5430 Add eth0 [10.128.3.49/23] from openshift-sdn
```

Talking about: `[ERROR] Failed to upload archive file for the build: quarkus-bot`.

Sgitario commented:
> So... I don't think it's the same issue but I still have an error when pushing the Quarkus bot to the production cluster (error which apparently doesn't prevent the deployment as I have a success in the end):
> 
> ```
> [INFO] [io.quarkus.container.image.openshift.deployment.OpenshiftProcessor] Applied: ImageStream ubi-quarkus-native-binary-s2i
> [INFO] [io.quarkus.container.image.openshift.deployment.OpenshiftProcessor] Applied: ImageStream quarkus-bot
> [INFO] [io.quarkus.container.image.openshift.deployment.OpenshiftProcessor] Applied: BuildConfig quarkus-bot
> [ERROR] Failed to upload archive file for the build: quarkus-bot
> [ERROR] Please check cluster events via `oc get events` to see what could have possibly gone wrong
> [WARNING] [io.quarkus.container.image.openshift.deployment.OpenshiftProcessor] An exception: 'Can't instantiate binary build, due to error reading/writing stream. Can be caused if the output stream was closed by the server.See if something's wrong in recent events in Cluster = Scheduled quarkus-bot-1-build.17489735390326d6 Successfully assigned prod-quarkus-bot/quarkus-bot-1-build to ip-10-0-142-232.us-west-2.compute.internal by ip-10-0-182-138
> Created quarkus-bot-1-build.17489735b5f74d2d Created container git-clone
> Started quarkus-bot-1-build.17489735b86fbe3a Started container git-clone
> BuildStarted quarkus-bot-1.17489735e85618ad Build prod-quarkus-bot/quarkus-bot-1 is now running
> AddedInterface quarkus-bot-1-build.17489735a33f5430 Add eth0 [10.128.3.49/23] from openshift-sdn
> ```
> 
> Talking about: `[ERROR] Failed to upload archive file for the build: quarkus-bot`.

I've also seen this issue even without this change. My guess is that the client is closing the connection after some time. How long did the build take for you? I would like to address this issue as part of this pull request as well, so I'll investigate it.

gsmet commented:
The build is a native one so it's a bit long, something like 2-3 minutes.

Sgitario commented:
After having tried up to 10 times a Native build to be deployed into OpenShift, I could reproduce the latter issue only once, and seems to be caused because OpenShift rejects the compressed file:

```
error: unable to extract binary build input, must be a zip, tar, or gzipped tar, or specified as a file: exit status 1 
```

And I think this is caused by `Files.move`(see [here](https://github.com/quarkusio/quarkus/pull/31503/commits/401fc01fd731214048881b36681688f4cefe7fec#diff-c15bcfcc919c2e97bb2cb66a3a836eb0537fbfb3a3fcfbd3373ad5df09749b63R368)). It moves files but depending on your environment, they might not be fully ready yet and hence OpenShift can't read it. Adding the `ATOMIC_MOVE` flag should address this issue.

With this change, I've tried the deployment to OpenShift and I could not reproduce the issue any longer.

@gsmet can you try again?

gsmet commented:
Sure. I’m at the doctor right now but will have a look when I’m back. 

The System.out.println should probably go away though :)

Sgitario commented:
> Sure. I’m at the doctor right now but will have a look when I’m back.
> 
> The System.out.println should probably go away though :)

Ops! I could reproduce the same failure even with this change, so it seems that it was not fixed :/

Sgitario commented:
@gsmet I've just realized that there is already an issue reported about the latter failure: https://github.com/quarkusio/quarkus/issues/31228 which is fixed in Kubernetes Client by https://github.com/fabric8io/kubernetes-client/pull/4899 in 6.5.0. 

I've just dropped the second commit, so we can merge this if you agree with.

gsmet commented:
@Sgitario yeah I totally agree we can merge it as is, it's already a big improvement.

As for the Kubernetes Client, do you know when they plan to release the fix? Just to know if it's going to be OK for Quarkus 3.0.

manusa commented:
> And the problem is that every time we invoke any of these methods, Fabric8 KubernetesClient tries to locate one HttpClient.Factory and it seems that the logic to get one HttpClient.Factory sometimes gets the VertxHttpClientFactory implementation over the expected one QuarkusHttpClientFactory.

This is kind of messed up, have you identified the root cause?


From the changes, I understand that the explicit HttpClient Factory is only used at build time, The runtime Bean will continue to be created by using the SPI-provided factory, right?

Sgitario commented:
> > And the problem is that every time we invoke any of these methods, Fabric8 KubernetesClient tries to locate one HttpClient.Factory and it seems that the logic to get one HttpClient.Factory sometimes gets the VertxHttpClientFactory implementation over the expected one QuarkusHttpClientFactory.

Not really, the logic I mentioned is [this one](https://github.com/fabric8io/kubernetes-client/blob/dfa920effaeef61909f2fac64d90ef676db93a77/kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/utils/HttpClientUtils.java#L151) (Note that there is an extra parenthesis in the third condition, but I'm not sure if this was intentional, however this was not the issue). Maybe, the root problem is how to load the factories [here](https://github.com/fabric8io/kubernetes-client/blob/dfa920effaeef61909f2fac64d90ef676db93a77/kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/utils/HttpClientUtils.java#L138). 

Anyway, I thought it was better to not call this logic at all to be more efficient and directly use the factory we want at build time. 

> This is kind of messed up, have you identified the root cause?
> 
> From the changes, I understand that the explicit HttpClient Factory is only used at build time, The runtime Bean will continue to be created by using the SPI-provided factory, right?

Yes, this is meant to be used only at build time. At runtime, the KubernetesClient instance will still be produced by https://github.com/quarkusio/quarkus/blob/main/extensions/kubernetes-client/runtime/src/main/java/io/quarkus/kubernetes/client/runtime/KubernetesClientProducer.java#L21.

Sgitario committed:
Fix randomly failures when deploying to OpenShift/K8s on JVM/Native

There were several places where a new KubernetesClient was being created:
- Some places use `Clients.fromConfig`.
- Other places use `KubernetesClientUtils.createConfig`.
- And other places use `KubernetesClientBuildItem.getClient`

And the problem is that every time we invoke any of these methods, Fabric8 KubernetesClient tries to locate one HttpClient.Factory and it seems that the logic to get one HttpClient.Factory sometimes gets the `VertxHttpClientFactory` implementation over the expected one `QuarkusHttpClientFactory`

With this solution, it avoids to find a HttpClient.Factory using the ServiceLoader logic in Fabric8 Kubernetes Client, but it provides the expected `QuarkusHttpClientFactory` implementation always.

Fix https://github.com/quarkusio/quarkus/issues/31476

Sgitario commented:
PR updated because I forgot to remove the file which after these changes is no longer needed: https://github.com/quarkusio/quarkus/pull/31503/files#diff-19ff5f9f5bb70cfddc9f0388d562c8b9ef15e74ead765a18f2ca00c4b99b1faa

yrodiere commented:
Merged, thanks!

gsmet commented:
> Yes, this is meant to be used only at build time. At runtime, the KubernetesClient instance will still be produced by 

@Sgitario but in this case, don't we have the same problem as having the two factories to compete while we want the Quarkus one to be used?

Sgitario commented:
> > Yes, this is meant to be used only at build time. At runtime, the KubernetesClient instance will still be produced by
> 
> @Sgitario but in this case, don't we have the same problem of having the two factories to compete while we want the Quarkus one to be used at runtime?

If this was not working at runtime either, then we should amend the producer to also use our custom HttpClient.Factory. Can you clarify @manusa or @geoand ?

manusa commented:
> Note that there is an extra parenthesis in the third condition, but I'm not sure if this was intentional, however this was not the issue

:facepalm: OK, I think I messed up in that one. The last condition looks definitely wrong.


> @Sgitario but in this case, don't we have the same problem of having the two factories to compete while we want the Quarkus one to be used at runtime?

One of the topics I discussed with @geoand while implementing the Quarkus factory was still providing the option for users to override the runtime HttpClient implementation. That's why I was specifically asking this. The Factory scoring system logic in the client impl needs to be fixed, then everything should work as expected.

manusa commented:
Sending the fix to the Client now.

Sgitario commented:
So, for runtime what should be the right behavior? Use the QuarkusHttpClientFactory impl by default? If used, I have to include back [the service binding file](https://github.com/quarkusio/quarkus/blob/9579d9d45df54f2a74eab3b9dadc09de50a763f7/extensions/kubernetes-client/runtime-internal/src/main/resources/META-INF/services/io.fabric8.kubernetes.client.http.HttpClient%24Factory).

manusa commented:
> So, for runtime what should be the right behavior? Use the QuarkusHttpClientFactory impl by default? If used, I have to include back [the service binding file](https://github.com/quarkusio/quarkus/blob/9579d9d45df54f2a74eab3b9dadc09de50a763f7/extensions/kubernetes-client/runtime-internal/src/main/resources/META-INF/services/io.fabric8.kubernetes.client.http.HttpClient%24Factory).

As I see it:
The QuarkusHttpClientFactory should score one point higher than the Vert.x factory provided by the client.
If a user wants to provide their own factory, it should score  higher than any other factory.
If the scoring system works as expected (and there aren't classloader issues), then the highest scoring non-default factory should be selected.

Sgitario commented:
So, for runtime what should be the right behavior? Use the QuarkusHttpClientFactory impl by default? 

If so, I'm not sure if I should include back [the service binding file](https://github.com/quarkusio/quarkus/blob/9579d9d45df54f2a74eab3b9dadc09de50a763f7/extensions/kubernetes-client/runtime-internal/src/main/resources/META-INF/services/io.fabric8.kubernetes.client.http.HttpClient%24Factory). 
But not sure it was needed it at all because we were already registering the service on [this line](https://github.com/quarkusio/quarkus/blob/9579d9d45df54f2a74eab3b9dadc09de50a763f7/extensions/kubernetes-client/deployment/src/main/java/io/quarkus/kubernetes/client/deployment/KubernetesClientProcessor.java#L229). I would need to test this.

manusa commented:
> But not sure it was needed it at all because we were already registering the service on [this line](https://github.com/quarkusio/quarkus/blob/9579d9d45df54f2a74eab3b9dadc09de50a763f7/extensions/kubernetes-client/deployment/src/main/java/io/quarkus/kubernetes/client/deployment/KubernetesClientProcessor.java#L229). I would need to test this.

AFAIU you need both. The file is used to register the SPI class. The second is used by Quarkus to make use of the SPI providers.

gsmet commented:
@Sgitario and I think you probably need to exclude the service binding file coming from the Kubernetes Client Vert.x jar to make sure the Quarkus one is the only one seen at runtime.

gsmet commented:
(by using `RemovedResourceBuildItem` to remove the service file from the jar)

Sgitario commented:
I've been playing with the `quarkus-kubernetes-client` extension and:

- The original issue https://github.com/quarkusio/quarkus/issues/31476 does not affect the KubernetesClient instance at runtime because the Vert.x extension is already disabling the DNS verification (here)[https://github.com/quarkusio/quarkus/blob/main/extensions/vertx/runtime/src/main/java/io/quarkus/vertx/core/runtime/VertxCoreRecorder.java#L300]. (I tried to reproduce the issue replicating what the kubernetes extension was doing and I did check that this flag was being enabled by the vertx extension).

- As @manusa said, as I removed the service file, the QuarkusHttpClientFactory implementation is not found at runtime, but the VertxHttpClientFactory impl, However, VertxHttpClientFactory works already fine because of the previous point. 

Therefore, I think we're ok with using the VertxHttpClientFactory impl at runtime and only QuarkusHttpClientFactory at build time, and we need no more changes. Wdyt?

Note that I'm seeing the following warning when using the Kubernetes Client:

```
2023-03-03 12:19:48,711 WARN  [io.ver.cor.imp.VertxImpl] (executor-thread-0) You're already on a Vert.x context, are you sure you want to create a new Vertx instance?
```

But this is unrelated to my changes.

manusa commented:
> Therefore, I think we're ok with using the VertxHttpClientFactory impl at runtime and only QuarkusHttpClientFactory at build time. So we don't need more changes. Wdyt?

Besides disabling the DNS verification, the purpose of the factory is to reuse the Vertx object instance at runtime which is why it was originally added.

https://github.com/quarkusio/quarkus/blob/f14cb409a7c78f7df913955457e3a7964993022f/extensions/kubernetes-client/runtime-internal/src/main/java/io/quarkus/kubernetes/client/runtime/QuarkusHttpClientFactory.java#L24

Sgitario commented:
I see. So, if we want to let users to overwrite it via ServiceLoader and maintain QuarkusHttpClientFactory, we definitely need to exclude the VertxHttpClientFactory impl using RemovedResourceBuildItem as @gsmet .
Let me prepare a pull request with this.

gsmet commented:
Thanks @Sgitario and @manusa !

Sgitario commented:
Fix: https://github.com/quarkusio/quarkus/pull/31582
